docs/scripts/preprocessor.py

```diff
@@ -0,0 +1,151 @@
+#!/usr/bin/python3
+
+"""
+preprocessor for sphinx files
+- reads in all .rst files in input directory
+- writes processed .rst files to output directory
+
+by copying files to output directory, we aren't
+modifying the  original .rst files in any way.
+
+preprocessor steps:
+- format tables
+"""
+
+import argparse
+import shutil
+import os
+
+
+def find_markdown_tables(lines):
+    """
+    return (row_start, row_end) for all markdown tables.
+    """
+    tpos = []
+    md_table_idx = None
+    for idx, line in enumerate(lines):
+        if line.startswith("|"):
+            if not md_table_idx:
+                md_table_idx = idx
+        else: # end of md table
+            if md_table_idx is not None:
+                tpos.append((md_table_idx, idx - 1))
+                md_table_idx = None
+    if md_table_idx is not None:
+        tpos.append((md_table_idx, idx))
+    return tpos
+
+
+def get_markdown_row_data(row):
+    """
+    get row (or header) data from markdown row (or header)
+
+    row
+        | apples | red | a |
+
+    data
+        ['apples', 'red', 'a']
+    """
+    data = row.split("|")
+    data = data[0:len(data)-1]
+    data = [e.strip() for e in data if e.strip()]
+    return data
```

Comment:
```suggestion
    return [e.strip() for e in row.split("|")][1:-1]
```

Reply:
This is an example of [ITM](https://www.youtube.com/watch?v=CjHgL5EQdcY). Are you able to try and refactor other places in the code do be more declarative?

Reply:
Sure will do tomorrow. Thanks for the suggestion.

Reply:
I have refactored the code. I am pretty certain there are no more uses of ITM anymore. Hope this falls under the category of "beautiful" python refactoring :)

https://github.com/codereport/jsource/pull/129/commits/9b5356b596555b8730407fee8c94f5af29c8a27c